contracts/contracts/strategies/CurveAMOStrategy.sol

```diff
@@ -0,0 +1,702 @@
+     * the pool's balance is not worsened.
+     * Withdrawals are proportional so doesn't change the pools asset balance.
+     */
+    modifier improvePoolBalance() {
+        // Get the hard asset and OToken balances in the Curve pool
+        uint256[] memory balancesBefore = curvePool.get_balances();
+        // diff = hardAsset balance - OTOKEN balance
+        int256 diffBefore = (
+            balancesBefore[hardAssetCoinIndex].scaleBy(
+                decimalsOToken,
+                decimalsHardAsset
+            )
+        ).toInt256() - balancesBefore[otokenCoinIndex].toInt256();
+
+        _;
+
+        // Get the hard asset and OToken balances in the Curve pool
+        uint256[] memory balancesAfter = curvePool.get_balances();
+        // diff = hardAsset balance - OTOKEN balance
+        int256 diffAfter = (
+            balancesAfter[hardAssetCoinIndex].scaleBy(
+                decimalsOToken,
+                decimalsHardAsset
+            )
+        ).toInt256() - balancesAfter[otokenCoinIndex].toInt256();
+
+        if (diffBefore == 0) {
+            require(diffAfter == 0, "Position balance is worsened");
+        } else if (diffBefore < 0) {
+            // If the pool was originally imbalanced in favor of OTOKEN, then
+            // we want to check that the pool is now more balanced
+            require(diffAfter <= 0, "OTokens overshot peg");
+            require(diffBefore < diffAfter, "OTokens balance worse");
+        } else if (diffBefore > 0) {
+            // If the pool was originally imbalanced in favor of hardAsset, then
+            // we want to check that the pool is now more balanced
+            require(diffAfter >= 0, "Assets overshot peg");
+            require(diffAfter < diffBefore, "Assets balance worse");
+        }
+    }
+
+    constructor(
+        BaseStrategyConfig memory _baseConfig,
+        address _otoken,
+        address _hardAsset,
+        address _gauge,
+        address _minter
+    ) InitializableAbstractStrategy(_baseConfig) {
+        lpToken = IERC20(_baseConfig.platformAddress);
+        curvePool = ICurveStableSwapNG(_baseConfig.platformAddress);
+        minter = ICurveMinter(_minter);
+
+        oToken = IERC20(_otoken);
+        hardAsset = IERC20(_hardAsset);
+        gauge = ICurveLiquidityGaugeV6(_gauge);
+        decimalsHardAsset = IBasicToken(_hardAsset).decimals();
+        decimalsOToken = IBasicToken(_otoken).decimals();
+
+        (hardAssetCoinIndex, otokenCoinIndex) = curvePool.coins(0) == _hardAsset
+            ? (0, 1)
+            : (1, 0);
+        require(
+            curvePool.coins(otokenCoinIndex) == _otoken,
+            "Invalid coin indexes"
+        );
+    }
+
+    /**
+     * Initializer for setting up strategy internal state. This overrides the
+     * InitializableAbstractStrategy initializer as Curve strategies don't fit
+     * well within that abstraction.
+     */
```

Comment:
Can we add Natspec for the params
```
     * @param _rewardTokenAddresses Address of CRV
     * @param _maxSlippage Maximum slippage allowed for adding/removing liquidity from the Curve pool.
```

Reply:
[9e86aa3](https://github.com/OriginProtocol/origin-dollar/pull/2436/commits/9e86aa3f272416bcc8242d9a4b3e1e586d3ee076)